### PR TITLE
Fixes a 2D triangle intersection case

### LINKED ISSUE #152: Failing edge case for 3D triangle intersection.
To recreate the issue, add the following to the `3D_triangle_triangle_intersection` test inside `primal_intersect.cpp` :

```
  tri3A = Triangle3( Point3::make_point(76.648, 54.6752, 15.0012),
                     Point3::make_point(76.648, 54.6752, 14.5542),
                     Point3::make_point(76.582, 54.6752, 14.7879) );
  
  tri3B = Triangle3( Point3::make_point(76.6252, 54.6752, 14.892),
                     Point3::make_point(76.582, 54.6752, 14.7879),
                     Point3::make_point(76.5617,54.6752,14.7929) );

  permuteCornersTest(tri3A, tri3B, "3D tris with diff permutation results",
                     false, true);
```

The test case fails with the following message:
```
Failed
Triangles should intersect but did not
	 -- 6 of 36 permutations failed when testing intersection of triangles 
	 a: {(76.647999999999996,54.675199999999997,15.001200000000001) (76.647999999999996,54.675199999999997,14.5542) (76.581999999999994,54.675199999999997,14.7879)}
	 b: {(76.625200000000007,54.675199999999997,14.891999999999999) (76.581999999999994,54.675199999999997,14.7879) (76.561700000000002,54.675199999999997,14.792899999999999)}
```

6 of the 36 permutations of the triangle intersection return `false` indicating there was no intersection, 30 of the 36 permutations of the triangle intersection return `true` indicating there was intersection.
Expected result is all permutations of the triangles should return the same value.

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- Fixes #152 
- There was a bug in one of the 2D triangle-triangle intersection cases in the original paper
- I added the test case from #152 and several variations
- I also added a number of 2D triangle-triangle tests where one of the points of the second triangle are known to be inside the first triangle.  It uses points from barycentric coordinates to ensure that we can get exact edge and corner cases where the other points of the triangles are on edges and corners :)
- Minor: I ran uncrustify on primal

### CODE REVIEW

File: src/axom/primal/operators/detail/intersect_impl.cpp [hunk 1]
```diff
@@ -330,35 +330,42 @@ inline bool checkVertex(const Point2& p1,
...
-          return false;
+          return false;                                                                      // q1 in in R_22 -- no intersection
         }
       }
-      else
+      else                                                                             // q1 is in R_24
       {
         if (isGpeq(twoDcross(p1, p2, r1), 0.0, includeBoundary, EPS) &&
-            isGpeq(twoDcross(r2, p2, r1), 0.0, includeBoundary, EPS))
+            isLpeq(twoDcross(q1, p2, r1), 0.0, includeBoundary, EPS))
```
Comment (kennyweiss) [inline src/axom/primal/operators/detail/intersect_impl.cpp:358]: The bug fix is clearly here:
```diff
- isGpeq(twoDcross(r2, p2, r1), 0.0, includeBoundary, EPS))   // Interpretation: check for CCW orientation (r2,p2,r1)
+ isLpeq(twoDcross(q1, p2, r1), 0.0, includeBoundary, EPS))   // Interpretation: check for CW orientation (q1,p2,r1)
```

Comment (gzagaris) [reply src/axom/primal/operators/detail/intersect_impl.cpp:358]: Nice find!

Comment (kennyweiss) [reply src/axom/primal/operators/detail/intersect_impl.cpp:358]: The bug was here:
image

File: src/axom/primal/operators/detail/intersect_impl.cpp [hunk 2]
```diff
@@ -330,35 +330,42 @@ inline bool checkVertex(const Point2& p1,
                         bool includeBoundary,
                         double EPS)
 {
-  if (isGpeq(twoDcross(r2, p2, q1), 0.0, includeBoundary, EPS))
+  // The tests `isGpeq(twoDcross(...))` are checking the orientation
+  // of the triangle defined by its three arguments (CCW vs CW)
+  // Note: Comments in this function refer to regions
+  // in Figure 8 of the paper: https://hal.inria.fr/inria-00072100/document
+
+  /* *INDENT-OFF* */
```
Comment (gzagaris) [inline src/axom/primal/operators/detail/intersect_impl.cpp:338]: Why do you disable the style checks here?

Comment (kennyweiss) [reply src/axom/primal/operators/detail/intersect_impl.cpp:338]: I added annotations to interpret the cases w.r.t. the figure in the paper and they would not fit in 80 chars.

Here is the legend:
image

File: src/axom/primal/operators/detail/intersect_impl.cpp [hunk 3]
```diff
@@ -330,35 +330,42 @@ inline bool checkVertex(const Point2& p1,
...
+
+  if (isGpeq(twoDcross(r2, p2, q1), 0.0, includeBoundary, EPS))                // q1 is in {R_22, R_23, R_24, R_25}
```
Comment (gzagaris) [inline src/axom/primal/operators/detail/intersect_impl.cpp:340]: The comments here are very helpful, but I think it would be easier to read if those comments were in the 1st line inside the if-block. It would also better conform with our style, keep lines at 80 characters, this block of code seems that is not conforming to that.

File: src/axom/primal/tests/primal_intersect.cpp [hunk 4]
```diff
@@ -76,7 +79,9 @@ void permuteCornersTest(const primal::Triangle< double, DIM > & a,
   {
     for (int j = 0 ; j < 3 ; ++j)
     {
-      if(primal::intersect(roll(a, i), roll(b, j),includeBdry) != testtrue)
+      auto t1 = roll(a, i);
+      auto t2 = roll(b, j);
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_intersect.cpp:83]: Extracting these variables made it much easier to debug the failing cases.

File: src/axom/primal/operators/detail/intersect_impl.cpp (hunk 1 above)
Comment (kennyweiss) [reply src/axom/primal/operators/detail/intersect_impl.cpp:358]: The case that helped me debug this case was:
```
Triangle 1 { p: (1,1),        q: (0,0),  r: (1,0)
Triangle 2 { p: (.667, .333), q: (1,-2), r: (1,-1)
```
It is in the new 2D barycentric triangle-triangle interesection tests.

File: src/axom/primal/operators/detail/intersect_impl.cpp (hunk 2 above)
Comment (agcapps) [reply src/axom/primal/operators/detail/intersect_impl.cpp:338]: Thanks for the `INDENT-OFF` annotations.

File: src/axom/primal/operators/detail/intersect_impl.cpp (hunk 3 above)
Comment (agcapps) [reply src/axom/primal/operators/detail/intersect_impl.cpp:340]: I appreciate the labels out to the side, so they're not squashed in among the code.  The style must allow exceptions to the 80-character rule to serve code legibility, rather than rigidly hindering it.

Comment (gzagaris) [reply src/axom/primal/operators/detail/intersect_impl.cpp:340]: @agcapps -- exceptions are allowed with the `INDENT-OFF`. 

However, we should be careful not to abuse this "feature". Littering the code with exceptions defeats the purpose of having a style in the first place. 

I prefer reading the case labels in the beginning of the code block. But, that is subjective.

File: src/axom/primal/tests/primal_intersect.cpp [hunk 5]
```diff
@@ -1073,7 +1073,7 @@ TEST( primal_intersect, 3D_triangle_triangle_intersection_regression )
   {
     // https://github.com/LLNL/axom/issues/152
     std::string msg = "From Axom github issue #152 (simplified)";
-    Point3 vdata[] = { Point3::make_point( 0.066,  0,  0.2133).
+    Point3 vdata[] = { Point3::make_point( 0.066,  0,  0.2133),
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_intersect.cpp:1076]: I found the bug in my test -- I had a period (`.`) instead of a comma (`,`) in this array initializer list.

Interestingly, the `toss3` tests all passed, but the `blueos` tests all failed in the previous version.

**Comment by kennyweiss** [comment]
Passed all LC host-configs on `toss3` and `blueos`
